### ISSUE TITLE
New Tag / Issues deploying on Google Cloud Run with HTTP/2 TLS

### ISSUE THREAD

**Issue by jgavris** [znly/bazel-cache#5]
I was very excited to try this out, as I've written a few simple REST bazel cache services (redis and S3 backends). Currently we use the REST backend directly to GCS and are hoping to move to gRPC for compression and latency improvements. I tried out the currently tagged version on the docker registry and wasn't able to get our workspace to connect. I've been debugging it on a branch here if you're curious.

https://github.com/znly/bazel-cache/compare/master...ButterflyNetwork:upgrades?expand=1

- Are you considering publishing a new tagged image?
- Do you have HTTP/2 over TLS working (for `grpcs://`) on Google Cloud Run? I'm using the 'v2' infra for GCR, and my YAML config shows this (using `--use-http2` for `gcloud deploy`), `h2c` being HTTP/2 cleartext.

```yaml
        ports:
        - name: h2c
          containerPort: 9092
```

cc @steeve @Thooms @SaveTheRbtz

**Comment by EliSchleifer** [comment]
@jgavris - did you deploy this container somewhere I can reference. Would like to try out this build

**Comment by jgavris** [comment]
@EliSchleifer I pushed it to our private container registry on GCS. You are more than welcome to check out my PR / branch and `make image` (you'll need Docker Buildx installed in addition to Docker). 

https://docs.docker.com/buildx/working-with-buildx/

I can send a cleaned up version of my `deploy.sh` but it is not much more than what's in the documentation here https://cloud.google.com/container-registry/docs/pushing-and-pulling and here https://cloud.google.com/run/docs/deploying, plus a little bit of magic to introspect the image hash.

**Comment by EliSchleifer** [comment]
I got the image built in my artifact repository. Would love to see what your deploy.sh looks like. I am hitting problems getting this launched cleanly.
________________________________

**Comment by jgavris** [comment]
I haven't fully cleaned this up, but you need to use `-use-http2`. If you can send a redacted YAML config from your cloud run service that might help debug it. Go to your service details and the YAML tab.

```sh
make image 
docker tag znly/bazel-cache:0.1.4 gcr.io/some-project/znly/bazel-cache:0.1.4
docker push gcr.io/some-project/znly/bazel-cache:0.1.4
export IMAGE=$(docker inspect --format='{{index .RepoDigests 0}}' gcr.io/some-project/znly/bazel-cache:0.1.4)
gcloud run deploy bazel-cache \
    --image=$IMAGE \
    --concurrency=10 \
    --platform=managed \
    --region=some-region \
    --cpu=4 \
    --memory=8Gi \
    --use-http2 \
    --set-env-vars "CACHE_URI=gcs://some-bucket?ttl_days=30" \
    --project=some-project
```

**Comment by jgavris** [comment]
Leave blank 'container command' and 'container arguments' as I made `serve` the default command in this image.

**Comment by EliSchleifer** [comment]
spec:
  template:
    metadata:
      name: bazel-cache-XXXXXX
      annotations:
        run.googleapis.com/client-name: cloud-console
        client.knative.dev/user-image: us-west1-docker.pkg.dev/XXXXXXX/bazel-cache/0.1.3@sha256:XXXXXXXXXX
        autoscaling.knative.dev/minScale: '1'
        run.googleapis.com/execution-environment: gen2
        autoscaling.knative.dev/maxScale: '100'
    spec:
      containerConcurrency: 80
      timeoutSeconds: 300
      serviceAccountName: XXXXXXX@XXXXXXXX.iam.gserviceaccount.com
      containers:
      - image: us-west1-docker.pkg.dev/XXXXXXX/bazel-cache/0.1.3@sha256:XXXXXXX
        ports:
        - name: h2c
          containerPort: 9092
        env:
        - name: CACHE_URI
          value: gcs://yyyyy-xxxx-zzzz
        resources:
          limits:
            cpu: 4000m
            memory: 8Gi
  traffic:
  - percent: 100
    latestRevision: true

**Comment by EliSchleifer** [comment]
Service is running now - but I hit this when using it:
build --remote_cache=https://bazel-run-XXXXXXX-uw.a.run.app

WARNING: Remote Cache: 3 errors during bulk transfer:
com.google.devtools.build.lib.remote.http.HttpException: 404 Not Found

**Comment by jgavris** [comment]
Your build command needs to include the `--remote_header=<identity_token>` if you didn't already do that. I'd expect an HTTP 401 though, not a 404. Did you grab an identity token for the user / IAM role that has access to the backing storage bucket?

https://github.com/znly/bazel-cache#authentication

**Comment by steeve** [comment]
Sorry I missed this as I was on paternity leave.

> I tried out the currently tagged version on the docker registry and wasn't able to get our workspace to connect. I've been debugging it on a branch here if you're curious.

Weird, we are running the latest tagged version IIRC.

> * Are you considering publishing a new tagged image?

Sure, I plan to add compression once I get back to work, but for now there's nothing to add?
My hunch is that compression will improve things massively wrt to latency.

> * Do you have HTTP/2 over TLS working (for `grpcs://`) on Google Cloud Run? I'm using the 'v2' infra for GCR, and my YAML config shows this (using `--use-http2` for `gcloud deploy`), `h2c` being HTTP/2 cleartext.

Yes, `grpcs://` is our current deployment strategy. That said, don't forget the `--remote_header` or else you'll get some permission issues, but it seems you got it figured out.

**Comment by EliSchleifer** [comment]
I redeployed to no avail here. Tried a couple things:
1) Brought in the full gcloud python library and wrap around bazel call. Previuosly I had set the service to no-auth just to diagnose it was working. Was still seeing 404 on all the GET/PUT calls into the service.
2) Then I noticed the code here is using "gcs://" url schema but it seems that googles own API specified "gs://" when I try gs:// I can an error bad url schema. So maybe that's just a red herring. I am not versed enough on the url schemes here and how Google exposes these.

Anything from this report shake a next step free in someones head?

**Comment by EliSchleifer** [comment]
Update: Got it working. Apparently the image i am using only works with grpc:// instead of the https I was using. Cache is working now as expected. Will want to do some analysis to figure out how this compares to just using GCP bucket like we were doing. I have to imagine it is much better

**Comment by jgavris** [comment]
You should be able to (and want to) use `grpcs://` in your bazel config with this image, that is what this issue was opened for and my PR implements. The benefit of gRPC is that it uses HTTP 2 which opens a single connection and pipelines requests (lower latency / round trips avoiding TLS handshakes).

**Comment by EliSchleifer** [comment]
yeah I realize that. Definitely superior protocol in this regard.

**Comment by jgavris** [comment]
OTOH, it's 'easy' to implement (gzip) compression over HTTP, most clients and servers can just [negotiate that](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding). For gRPC, we have to implement it in this server specifically on a per-message basis.
